### PR TITLE
[WIP] implement binary trie

### PR DESCRIPTION
**Pull request by jangko**
the goal of this PR are:
* implement working binary trie
* good test coverage for binary trie and it's components
* decent documentation of it's API and internal workings

### CODE REVIEW

**Comment by zah** [comment]
You've probably already noticed that the existing code in `./ethereum_trie` was translated from Python with https://github.com/metacraft-labs/py2nim.

Feel free to remove anything that is no longer useful. For the hexary trie, I ended up throwing away most of the translated code and created a new implementation.

File: ethereum_trie/utils/binaries.nim [hunk 1]
```diff
@@ -1,43 +1,77 @@
...
+  for i in 0..<8:
+    result[7 - i] = byte(1 shl i)
+
+const
+  byteMask = generateMask()
+  TWO_BITS = ["\x00\x00", "\x00\x01", "\x01\x00", "\x01\x01"]
+  PREFIX_00 = "\x00\x00"
+  PREFIX_100000 = "\x01\x00\x00\x00\x00\x00"
+
+proc encodeToBin*(value: BytesRange): Bytes =
```
Comment (zah) [inline ethereum_trie/utils/binaries.nim]: Isn't it possible to remove these memory allocating functions that store bits as bytes?

We can just implement helpers for accessing the individual bits of a `Bytes` values.

File: ethereum_trie/binary.nim [hunk 2]
```diff
@@ -1,158 +1,95 @@
...
+  result.dbLink = db
+  result.rootHash = BLANK_HASH.toTrieNodeKey
+
+proc `==`(a, b: BytesRange): bool =
+  if a.len != b.len: return false
+  for i in 0..<a.len:
+    if a[i] != b[i]: return false
+  result = true
+
+proc getAux[DB](self: BinaryTrie[DB], nodeHash: TrieNodeKey, keyPath: BinVector): BytesRange =
```
Comment (zah) [inline ethereum_trie/binary.nim]: Prefer the implicit generic style when you don't need to reference the generic parameters:

``` nim
proc getAux(self: BinaryTrie, nodeHash: TrieNodeKey, keyPath: BinVector): BytesRange = 
```

This makes the code shorter and more resilient to refactoring (we can add an remove generic parameters without changing all the procs defined in the module).

File: ethereum_trie/binary.nim [hunk 3]
```diff
@@ -1,158 +1,95 @@
...
+  assert hash.len == 32
+  for i in 0..<32:
+    result.hash.data[i] = uint8(hash[i])
+  result.usedBytes = 32
+
+proc initBinaryTrie*[DB](db: ref DB): BinaryTrie[DB] =
+  result.dbLink = db
+  result.rootHash = BLANK_HASH.toTrieNodeKey
+
+proc `==`(a, b: BytesRange): bool =
```
Comment (zah) [inline ethereum_trie/binary.nim]: Isn't this already provided by the nim-ranges lib?

File: ethereum_trie/binary.nim [hunk 4]
```diff
@@ -1,158 +1,95 @@
...
+  for i in 0..<a.len:
+    if a[i] != b[i]: return false
+  result = true
+
+proc getAux[DB](self: BinaryTrie[DB], nodeHash: TrieNodeKey, keyPath: BinVector): BytesRange =
+  # Empty trie
+  if nodeHash == BLANK_HASH:
+    return zeroBytesRange
+
+  let (nodeType, leftChild, rightChild) = parseNode(self.dbLink[].get(nodeHash.hash).toRange)
```
Comment (zah) [inline ethereum_trie/binary.nim]: Please use an accessor template like this:

``` nim
template db(t: BinaryTrie): untyped = t.dbLink[]
```

I plan to change how the `dbLink` field is stored in the Trie objects and the accessor template will be able to hide this detail.

File: ethereum_trie/nodes.nim [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    raise newException(InvalidNode, "Unable to parse node")
+
+proc encodeKVNode*(keyPath: Bytes, childNodeHash: Bytes): Bytes =
+  ## Serializes a key/value node
+
+  if keyPath.len == 0:
+    raise newException(ValidationError, "Key path can not be empty")
+  assert(childNodeHash.len == 32)
+
+  let encodedKey = encodeFromBinKeypath(keyPath.toRange)
```
Comment (zah) [inline ethereum_trie/nodes.nim]: This code performs too many allocations and copies. If the use of binary arrays is removed, it seems to me that a single `seq` allocation will be sufficient here and the byte fragments can be copied in their right places in-place without additional allocations.

File: tests/test_nodes_utils.nim [hunk 6]
```diff
@@ -56,4 +56,129 @@ proc testConsumeCommonPrefix*(left: seq[int]; right: seq[int];
...
+    except AssertionError as E:
+      check (E.msg == "len(leftChildNodeHash) == 32 ") or (E.msg == "len(rightChildNodeHash) == 32 ")
+    except:
+      doAssert(false)
+
+const
+  leafData = [
+    ("\x03\x04\x05", "\x02\x03\x04\x05"),
+    ("", None)
+  ]
```
Comment (zah) [inline tests/test_nodes_utils.nim]: Let's add a randomized test similar to this one:
https://github.com/ethereum/py-trie/blob/master/tests/test_bin_trie.py#L20

You can use the Nim's quickcheck library for this:
https://github.com/alehander42/nim-quicktest

**Comment by jangko** [comment]
TODO list:
[ ] replace keccak_tiny with NimCrypto, edit nimble file
[ ] replace encodeToBin with bit-vector-something
[ ] implicit generic style
[ ] combine encodeKVNode allocation to single alloc
[ ] use db accessor template
[x] remove equality operator
[ ] add random test for nodes utils + nim-quicktest
[ ] write binary trie test + nim-quicktest
[ ] write branch utils
[ ] write branch utils test
[ ] polish doc
[ ] add api doc
[ ] add example doc
[ ] source code file reorganization
[ ] optimization

**Comment by zah** [comment]
For the `BitVector`, you can either go with helper accessors (perhaps that's enough) or you can implement something like the `NibblesRange` used in the hexary trie:
https://github.com/status-im/nim-trie/blob/3fc5dcc67c2b7cfdf6ec9b9772da9da0e50cd5e7/ethereum_trie/nibbles.nim

Please note that the `BytesRange` objects are cheap to copy, because they store an underlying `shallow` sequence (so copying touches only one pointer and two ints).

**Comment by zah** [comment]
Merged into master, so I can start fixing the hexary trie code. You can keep pushing commits here until the feature is completely ready.

**Pull request merged**